### PR TITLE
Refactor ReportScreenWrapper to functional component

### PR DESCRIPTION
**Pull request by kosmydel**
### Details

This PR addresses two issues:
- Refactor ReportScreenWrapper to functional component,
- Explore using a skeleton for loading an app after sign in.

### Fixed Issues

Related to https://github.com/Expensify/App/pull/19263#discussion_r1220291991
$ https://github.com/Expensify/App/issues/20320

### Tests

**iOS/Android/mWeb iOS**:
1. Log in to the account.
2. Enter the report by selecting it from the list, and check if the report is displayed correctly.
3. Leave the report by pressing the back button.
4. Enter another report.
5. Leave the report.

**Web/desktop**:
1. Log in to the account.
2. Check if while loading the skeleton is displayed instead of the loading circle.
3. Check if the last report is opened.
4. Navigate between reports several times.
3 Navigate to [http://localhost:8080/r/](http://localhost:8080/r/) from the address bar.
5. Check if the last selected chat is opened.
6. Log out.
7. Create a new account.
8. Check if the user is forwarded to the `Concierge` report.

### Offline tests

Checked on the web version with "Slow 3G" option if the flow works.
### QA Steps

- [ ] Verify that no errors appear in the JS console
- [ ] Log in to the account with several reports, and navigate between reports.
- [ ] In the address bar type (or with a different prefix) `http://localhost:8080/r/`, and check if the last report is automatically selected and loaded.
- [ ] On the web/desktop: Create a new account, log in, and verify if the user can see the Concierge report.
- [ ] On the web/desktop:
  - Log in to the existing account.
  - Select a random report and remember which report you have chosen.
  - Log out from the account.
  - Log in again to the same account.
  - Check if the previously chosen report is displayed.
- [ ] On the web/desktop:
  - Log in to the account.
  - Check if the skeleton is displayed in the report screen while loading.
  - Check if the report is loaded successfully.

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/104823336/b3973bd0-83cb-43d8-a50f-f394e7a79fbd

**Previous versions:**

https://github.com/Expensify/App/assets/104823336/69341eaf-2d2f-4e3b-8cc4-d0049c70af97

https://github.com/Expensify/App/assets/104823336/2a0c9426-a2ab-4b19-bc41-207c1ea7a5b2

</details>

<details>
<summary>Mobile Web - Chrome</summary>

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/104823336/b145c223-5576-4b48-8f4f-2054403f7e75

**Previous versions:**

https://github.com/Expensify/App/assets/104823336/f1262ecc-d17a-4b74-a9ce-b0748f4cb6e0

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/104823336/e039da0f-6262-468a-b329-f94cae9114b7

**Previous versions:**

https://github.com/Expensify/App/assets/104823336/42e6fe75-910b-4d03-b5e9-5f732efb957d

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/104823336/a32bed44-40ef-4a0f-9970-452830895267

**Previous versions:**

https://github.com/Expensify/App/assets/104823336/9c2c0ea1-910d-446b-8988-ec8214c90be8

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/104823336/fd8c1fce-3615-4a50-b479-a44c55b2f3dc

**Previous versions:**

https://github.com/Expensify/App/assets/104823336/48acd8d4-a686-4ff0-877d-286f4391cffb

</details>

### CODE REVIEW

**Comment by kosmydel** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by kosmydel** [comment]
recheck

**Comment by mountiny** [comment]
I have bumped SWM i think @kosmydel is out for uni exams

File: src/libs/Navigation/AppNavigator/ReportScreenWrapper.js
[the remark this answers is not in the document]
Comment (kosmydel) [reply src/libs/Navigation/AppNavigator/ReportScreenWrapper.js:1]: Hey, sorry for the delay in the response.

I'm not sure what you mean by that. I couldn't find the `useBetas` hook in our codebase. Could you explain it a little more in detail?

- Where should I get from the `useBetas` hook?
- For what should I use it (instead of `betas` in `withOnyx`?)

**Comment by kosmydel** [comment]
Changed to draft, because I am going to add to this PR fix to a [related issue](https://github.com/Expensify/App/pull/19263#discussion_r1220296500) (using skeleton instead of `FullScreenLoadingIndicator`).

**Comment by kosmydel** [comment]
I've updated the PR description and added a fix for displaying the skeleton when logging in.

**Review by parasharrajat — CHANGES_REQUESTED** [review]

File: src/pages/home/ReportScreen.js [hunk 1]
```diff
@@ -113,7 +113,7 @@ const defaultProps = {
  * @returns {String}
  */
 function getReportID(route) {
-    return route.params.reportID.toString();
+    return lodashGet(route, 'params.reportID', null);
```
[the remark this answers is not in the document]
Comment (kosmydel) [reply src/pages/home/ReportScreen.js:116]: The main reason is to show the skeleton while the `reportID` is not loaded yet.
I made this change to allow `ReportScreen` to use `null` as `reportID`, which then is used to display the skeleton.
This approach was very convenient, because `ReportScreen` already implements skeleton. But, as you say, it may lead to some bugs with Onyx. I couldn't find in the docs option, to 'delay' the Onyx, until we have all required props.

Another solution might be, to add skeletons (when `reportID` is still loading) to the `ReportScreenWrapper`. If we want it to look exactly the same as it was before the navigation refactors, we would have to reimplement skeletons and add a header and footer there as well. It would be copying the same work as in `ReportScreen`, and would be poorly maintainable.

To sum up, I can:
- somehow use `ReportScreen` to display skeletons without the `reportID`,
- display the skeletons in the `ReportScreenWrapper`.

In my opinion, the first option would be cleaner and more maintainable. But I'm not sure if this is possible to work with `withOnyx`.

[the remark this answers is not in the document]
Comment (kosmydel) [reply src/pages/home/ReportScreen.js:117]: The main reason is to show the skeleton while the `reportID` is not loaded yet (in the wrapper).
I made this change to allow the  `ReportScreen` to use `null` as `reportID`, which then is used to display the skeleton.

This approach is very convenient because `ReportScreen` already implements a skeleton. But, as you say, it may lead to some bugs with Onyx. 

I created a stripped-down version of `ReportScreen` when we don't know the `reportID` yet. It is used in `ReportScreenWrapper`.

---

Another solution might be, to add skeletons (when reportID is still loading) to the ReportScreenWrapper. But I don't think the first approach is better.

**Review by parasharrajat — CHANGES_REQUESTED** [review]

**Comment by mountiny** [comment]
Thanks @parasharrajat lets get this one wrapped soon

**Review by parasharrajat — APPROVED** [review]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

:ribbon: :eyes: :ribbon:  C+ reviewed

File: src/pages/home/ReportScreen.js [hunk 2]
```diff
@@ -113,7 +113,7 @@ const defaultProps = {
  * @returns {String}
  */
 function getReportID(route) {
-    return route.params.reportID.toString();
+    return String(lodashGet(route, 'params.reportID', null));
```
Comment (mountiny) [inline src/pages/home/ReportScreen.js:116]: Should the default value be null here? shouldnt it be ''? If null then we need to update the returns docs too, right?

Comment (mountiny) [reply src/pages/home/ReportScreen.js:116]: Oh I completely missed that its still wrapped in string :D thanks

**Review by mountiny — APPROVED** [review]
Thanks @kosmydel and @parasharrajat
